src/uhs/atomizer/shard/controller.cpp

```diff
@@ -63,6 +63,19 @@ namespace cbdc::shard {
             return atomizer_handler(std::forward<decltype(pkt)>(pkt));
         });
 
+        constexpr auto max_wait = 3;
+        for(size_t i = 0; i < max_wait && m_shard.best_block_height() < 1;
+            i++) {
+            m_logger->info("Waiting to sync with atomizer");
+            constexpr auto wait_time = std::chrono::seconds(1);
+            std::this_thread::sleep_for(wait_time);
+        }
+
+        if(m_shard.best_block_height() < 1) {
```

Comment:
So, to clarify, this addition isn't enforcing the behavior of failing out (which will be hit by `ss` not having a value on now-line 79). Rather, this is adding the timeout to offer an alternative to immediately failing out if the synchronization hasn't already happened?

Reply:
Before this change the shard server (listening for sentinel requests) would start listening even if it was unable to connect to any atomizers. It would then start processing requests, all of which would silently fail, leading to needless errors /retries at the start of a test run. This change delays the start until we're definitely connected to the atomizer cluster (indicated by receiving at least one block), or a timeout elapses.

Reply:
Indeed. Eventually the shard might connect though.

If I remove the timeout, the integration tests fail. I think we'd need to rearchitect our tests to handle components starting in an asynchronous manner.

Reply:
Just to double-check I'm understanding: when you remove the timeout (i.e., it will wait indefinitely until it synchronizes), that results in the tests failing? That's not what I would expect (I'd have expected that to result in much more consistent behavior for testing).

Reply:
Indeed, if we remove the timeout the integration tests fail because the shard controller never returns from `init()`. Similarly, if we leave in the timeout but make failure to connect an error, `init()` returns false and the shard fails to initialize. This is because of the startup order in our integration tests: https://github.com/mit-dci/opencbdc-tx/blob/trunk/tests/integration/shard_integration_test.cpp#L25. `init()` must return before the test framework can connect to it to simulate messages.


e.g. making the timeout an error
```
[ RUN      ] shard_integration_test.error_initial_sync
[2022-06-03 11:18:03.521] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:18:04.524] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:18:05.528] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:18:06.533] [ERROR] Shard still not syncronized with atomizer
/build/tests/integration/shard_integration_test.cpp:24: Failure
Value of: m_ctl->init()
  Actual: false
Expected: true
[  FAILED  ] shard_integration_test.error_initial_sync (3148 ms)
```

e.g. removing the timeout
```
[ RUN      ] shard_integration_test.error_non_existant_input
[2022-06-03 11:26:37.088] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:38.093] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:39.098] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:40.100] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:41.102] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:42.105] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:43.109] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:44.112] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:45.118] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:46.118] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:47.120] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:48.123] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:49.128] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:50.133] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:51.136] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:52.139] [INFO ] Waiting to sync with atomizer
[2022-06-03 11:26:53.142] [INFO ] Waiting to sync with atomizer
...
```

Reply:
I see. That is an unfortunate coincidence of less-than-ideal behavior. I'm not against merging this, but I'm a little concerned about adding workarounds for the test-suite's fragility into the code.

Reply:
This is a bigger issue than can be reasonably dealt with in this PR. I've split out the larger issue to #127. I will think some more on whether or not it is sensible to merge in this work-around in the meantime (recognizing it's not needed to pass CI).